src/tests/cases/raw_mode.rs

```diff
@@ -105,7 +105,7 @@ fn multiple_packets_of_traffic_from_different_connections() {
             "2.2.2.2",
             "10.0.0.2",
             54321,
-            443,
+            4434,
```

Comment:
I'm not sure I understand the changes in the tests - could you explain them a little?

Reply:
In the old test cases, multiple processes are owning local port 443, but each process is connected to a different remote socket. This simulated scenario does not go well with the new logic of identifying processes, since we now expect a local port to be owned by a single process. So I changed the test cases so that each process is on a different port.

Reply:
> we now expect a local port to be owned by a single process.
I'm thinking out loud, correct me if I'm wrong, but can't you assign multiple IPs to a single network interface? In that case, can't two processes use the same local port for the different IPs?

Reply:
> but can't you assign multiple IPs to a single network interface

To the best of my knowledge, you can't. Different interfaces(WiFi/ethernet) can have different  IPs. Which means 1 process can use port 443 of WiFi, and another one can use port 443 of ethernet. When identifying processes, local_ip is used along with port number and protocol, so we should be able to handle this situation correctly.

Reply:
The `ip` command allows you to add multiple IPs to the same interface, but digging a bit into it I think it would create an "alias", so maybe the interface name would be different for our purposes, see [here](https://www.tecmint.com/create-multiple-ip-addresses-to-one-single-network-interface/) for example.

In any case, I think that's a niche case, not going to nitpick :) But I will try to setup this just to test how it works if I have time.

Reply:
Hrm, @zhangxp1998 - I see what you're saying about the tests. But take note: in this case for example, this is not the case. These are incoming packets directed at the same local server.

Reply:
> Hrm, @zhangxp1998 - I see what you're saying about the tests. But take note: in this case for example, this is not the case. These are incoming packets directed at the same local server.

I think that's what `one_process_with_multiple_connections` does? This test is multiple process multiple connections?

Reply:
Updated test cases

Reply:
You very much can assign multiple IPs to a single interface. In fact, you already have exactly that: 127.0.0.1 and 127.0.0.2. IPv4 localhost is almost every single 127.n.n.n (iirc 127.0.0.0 and 127.255.255.255 act strangely).

You can't do the same on IPv6, its localhost is a single address. Could probably assign a subnet, but don't quote me on that.

And yes, it is a niche case. It's worth considering, but not necessarily worth implementing.

Reply:
> You very much can assign multiple IPs to a single interface. In fact, you already have exactly that: 127.0.0.1 and 127.0.0.2. IPv4 localhost is almost every single 127.n.n.n (iirc 127.0.0.0 and 127.255.255.255 act strangely).
> 
> You can't do the same on IPv6, its localhost is a single address. Could probably assign a subnet, but don't quote me on that.
> 
> And yes, it is a niche case. It's worth considering, but not necessarily worth implementing.

Hmm can you make 1 interface have more than 1 external IPs?(how does routing/ISP work then?)

Reply:
I guess ARP works the same as before, but don't quote me on this

Reply:
@zhangxp1998 - in the tests, `10.0.0.2` is always the local IP (https://github.com/imsnif/bandwhich/blob/master/src/tests/fakes/fake_input.rs#L143)
When we build the packet, we first send the source and then the destination: https://github.com/imsnif/bandwhich/blob/master/src/tests/cases/ui.rs#L22-L38

Reply:
> Hmm can you make 1 interface have more than 1 external IPs?(how does routing/ISP work then?)

I also know that this is possible (eg. on routers - they cannot be limited by their physical interfaces, otherwise it would make things very difficult) - but I have a feeling @Alcaro is more in on the nitty-gritty details here than I am.

But indeed, very niche for our use-case.

Reply:
> > Hmm can you make 1 interface have more than 1 external IPs?(how does routing/ISP work then?)
> 
> I also know that this is possible (eg. on routers - they cannot be limited by their physical interfaces, otherwise it would make things very difficult) - but I have a feeling @Alcaro is more in on the nitty-gritty details here than I am.
> 
> But indeed, very niche for our use-case.

Routers? IIRC NAT boxes themselves typically have 1 IP? But they modify the packet's destination/source when forwarding packets. Not sure why non-NAT box routers need to deal with multiple IP on same interface though

Reply:
One example: https://learningnetwork.cisco.com/thread/21286

Reply:
> Hmm can you make 1 interface have more than 1 external IPs?

Does IPv4 plus IPv6 count?

If no, still possible, but a lot less common.

sudo ip addr add 192.168.3.1/28 dev enp1s0f0
sudo ip addr add 192.168.4.1/28 dev enp1s0f0

Pingable and bindable, and most likely externally reachable if I had a suitably configurable router. Can't get IPv6 working that way, though - they show up in ifconfig, but I get EADDRNOTAVAIL if I try to bind them. No idea why, networking is tricky.

> (how does routing/ISP work then?)

Routing/forwarding mostly stays in the kernel, it doesn't show up in lsof. Copying the packets to userspace would be a performance sinkhole. (Maintaining the NAT table may be partially in userspace, not sure.)

Reply:
> > Hmm can you make 1 interface have more than 1 external IPs?
> 
> Does IPv4 plus IPv6 count?
> 
> If no, still possible, but a lot less common.
> 
> sudo ip addr add 192.168.3.1/28 dev enp1s0f0
> sudo ip addr add 192.168.4.1/28 dev enp1s0f0
> 
> Pingable and bindable, and most likely externally reachable if I had a suitably configurable router. Can't get IPv6 working that way, though - they show up in ifconfig, but I get EADDRNOTAVAIL if I try to bind them. No idea why, networking is tricky.
> 
> > (how does routing/ISP work then?)
> 
> Routing/forwarding mostly stays in the kernel, it doesn't show up in lsof. Copying the packets to userspace would be a performance sinkhole. (Maintaining the NAT table may be partially in userspace, not sure.)

```
sudo ip addr add 192.168.3.1/28 dev enp1s0f0
sudo ip addr add 192.168.4.1/28 dev enp1s0f0
```
I suppose this only adds an additional IP behind the NAT box(AKA private IP)?

Reply:
Of course it's not globally accessible. If it was, I'd add 8.8.8.8 to my machine and see how much of the internet breaks.

I'm not even sure if it'd work in the LAN. I think it'd work on most routers, but probably not all - it depends on how they handle ARP requests. (The chosen addresses must be within the router's netmask, otherwise it'll send the packets in wrong direction.)

I can't test locally - the extra addresses I add don't stick. Either something (NetworkManager?) is instantly undoing my config, or ip addr silently fails. (My previous tests were done on an unplugged ethernet port.)

Of course such a configuration is extremely rare outside experimental conditions, even without unrelated processes sharing a port number. If supporting it would be troublesome, there's no need to.